File: Data Structures & Algorithms/search-for-word/submission-6.py
class Solution:

    def dfs(self, x:int,y:int, board:List[List[str]], word:str,index:int)->bool:
        
        if x<0 or y<0 or x>= len(board) or y>=len(board[0]):
            return False

        if board[x][y] != word[index]:
            return False

        if index == len(word)-1:
            return True

        temp = board[x][y]
        board[x][y] = "#"
        found = (
            self.dfs(x+1,y,board,word,index+1) or
            self.dfs(x,y+1,board,word,index+1) or
            self.dfs(x-1,y,board,word,index+1) or
            self.dfs(x,y-1,board,word,index+1)
         )

        board[x][y] = temp
        return found


    def exist(self, board: List[List[str]], word: str) -> bool:
        
        for i in range(len(board)):
            for j in range(len(board[0])):
                if self.dfs(i,j,board,word,0):
                    return True

        return False

File: Data Structures & Algorithms/permutation-string/submission-6.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int n = s1.size();
        int m = s2.size();
        vector<int>v1(26,0);
        vector<int>v2(26,0);
        if(n>m){
            return false;
        }

        for(int i=0;i<n;i++){
            v1[s1[i]-'a']++;
        }

        for(int i=0;i<n;i++){
            v2[s2[i]-'a']++;
        }

        if(v1==v2){
            return true;
        }
    //    int n = s1.size();
        for(int i=n;i<s2.size();i++){
            v2[s2[i]-'a']++;
            v2[s2[i-n]-'a']--;
            if(v1==v2){
                return true;
            }
        }
        return false;

    }
};
